### PR TITLE
make a bunch of math foldable

### PR DESCRIPTION
**Pull request by oscardssmith**
previously hyberbolic trig and `expm1` didn't fold because `ext_horner` couldn't be proved consistent or to terminate.

### CODE REVIEW

**Comment by ViralBShah** [comment]
Would be great to have some examples of potential gains.

**Comment by KristofferC** [comment]
Are these regressions on 1.8 at it is now?

File: base/math.jl [hunk 1]
```diff
@@ -293,7 +293,7 @@ end
 
 # polynomial evaluation using compensated summation.
 # much more accurate, especially when lo can be combined with other rounding errors
-@inline function exthorner(x, p::Tuple)
+Base.@assume_effects :total  @inline function exthorner(x::Union{Float16, Float32, Float64}, p::Tuple)
```
Comment (Keno) [inline base/math.jl:296]: Can we get away with weaker effects here (just :terminates) by replacing the tuple getindex by getfield?

**Comment by oscardssmith** [comment]
these are not regressions. there was no point in time where these were foldable.

File: base/math.jl (hunk 1 above)
Comment (oscardssmith) [reply base/math.jl:296]: yes, but imo, that's not noticably better.

Comment (Keno) [reply base/math.jl:296]: yes it is. `:total` is a very strong annotation that's easy to accidentally invalidate and cause segfaults.

**Comment by oscardssmith** [comment]
the gain is that sinh,cosh and expm1 now can be constant folded.

File: base/math.jl (hunk 1 above)
Comment (Keno) [reply base/math.jl:296]: #46134

Comment (oscardssmith) [reply base/math.jl:296]: hmm,
```
Base.@assume_effects :terminates_globally @noinline g(p::Tuple) 
    s=0
    for i in length(p):-1:1
        s+=getfield(p,i)
    end
    s
end
Base.infer_effects(g, (NTuple{5,Int},))
(!c,+e,!n,+t,+s)
```
Any idea what's going wrong here?

Comment (Keno) [reply base/math.jl:296]: Do you have #46044 merged?

Comment (oscardssmith) [reply base/math.jl:296]: good point.

File: base/math.jl [hunk 2]
```diff
@@ -293,10 +293,10 @@ end
 
 # polynomial evaluation using compensated summation.
 # much more accurate, especially when lo can be combined with other rounding errors
-@inline function exthorner(x, p::Tuple)
+Base.@assume_effects :terminates_locally  @inline function exthorner(x::Union{Float16, Float32, Float64}, p::Tuple)
```
Comment (Keno) [inline base/math.jl:296]: Why the extra type restriction?

**Review by Keno — APPROVED** [review]
LGTM

File: base/math.jl (hunk 2 above)
Comment (oscardssmith) [reply base/math.jl:296]: when I marked it `:total` I wanted to be slightly more cautions, but now that it's only `:terminates_locally` I'll relax it again.

**Comment by oscardssmith** [comment]
do we want to merge this?

File: base/math.jl [hunk 3]
```diff
@@ -293,10 +293,10 @@ end
 
 # polynomial evaluation using compensated summation.
 # much more accurate, especially when lo can be combined with other rounding errors
-@inline function exthorner(x, p::Tuple)
+Base.@assume_effects :terminates_locally  @inline function exthorner(x, p::Tuple)
```
Comment (aviatesk) [inline base/math.jl:296]:
```suggestion
Base.@assume_effects :terminates_locally @inline function exthorner(x, p::Tuple)
```

File: base/math.jl [hunk 4]
```diff
@@ -293,10 +293,10 @@ end
...
     hi, lo = p[end], zero(x)
     for i in length(p)-1:-1:1
-        pi = p[i]
+        pi = getfield(p, i)
```
Comment (aviatesk) [inline base/math.jl:299]: Add a comment to explain that this `getfield` is necessary to achieve `:consistent`-cy.

**Comment by oscardssmith** [comment]
bump

**Review by aviatesk — APPROVED** [review]

**Pull request merged**